sync2/tokens_table.go

```diff
@@ -0,0 +1,233 @@
+	// We cannot use bcrypt/scrypt as we need the plaintext to do sync requests!
+	key256 []byte
+}
+
+// NewTokensTable creates the syncv3_sync2_tokens table if it does not already exist.
+func NewTokensTable(db *sqlx.DB, secret string) *TokensTable {
+	db.MustExec(`
+	CREATE TABLE IF NOT EXISTS syncv3_sync2_tokens (
+		token_hash TEXT NOT NULL PRIMARY KEY, -- SHA256(access token)
+		token_encrypted TEXT NOT NULL,
+		-- TODO: FK constraints to devices table?
+		user_id TEXT NOT NULL,
+		device_id TEXT NOT NULL,
+		last_seen TIMESTAMP WITH TIME ZONE NOT NULL
+	);`)
+
+	// derive the key from the secret
+	hash := sha256.New()
+	hash.Write([]byte(secret))
+
+	return &TokensTable{
+		db:     db,
+		key256: hash.Sum(nil),
+	}
+}
+
+func (t *TokensTable) encrypt(token string) string {
+	block, err := aes.NewCipher(t.key256)
+	if err != nil {
+		panic("sync2.DevicesTable encrypt: " + err.Error())
+	}
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		panic("sync2.DevicesTable encrypt: " + err.Error())
+	}
+	nonce := make([]byte, gcm.NonceSize())
+	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
+		panic("sync2.DevicesTable encrypt: " + err.Error())
+	}
+	return hex.EncodeToString(nonce) + " " + hex.EncodeToString(gcm.Seal(nil, nonce, []byte(token), nil))
+}
+func (t *TokensTable) decrypt(nonceAndEncToken string) (string, error) {
+	segs := strings.Split(nonceAndEncToken, " ")
+	nonce := segs[0]
+	nonceBytes, err := hex.DecodeString(nonce)
+	if err != nil {
+		return "", fmt.Errorf("decrypt nonce: failed to decode hex: %s", err)
+	}
+	encToken := segs[1]
+	ciphertext, err := hex.DecodeString(encToken)
+	if err != nil {
+		return "", fmt.Errorf("decrypt token: failed to decode hex: %s", err)
+	}
+	block, err := aes.NewCipher(t.key256)
+	if err != nil {
+		return "", err
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", err
+	}
+	token, err := aesgcm.Open(nil, nonceBytes, ciphertext, nil)
+	if err != nil {
+		return "", err
+	}
+	return string(token), nil
+}
+
+func hashToken(accessToken string) string {
+	// important that this is a cryptographically secure hash function to prevent
+	// preimage attacks where Eve can use a fake token to hash to an existing device ID
+	// on the server.
+	hash := sha256.New()
+	hash.Write([]byte(accessToken))
+	return hex.EncodeToString(hash.Sum(nil))
+}
+
+// Token retrieves a tokens row from the database if it exists.
+// Errors with sql.NoRowsError if the token does not exist.
+// Errors with an unspecified error otherwise.
+func (t *TokensTable) Token(plaintextToken string) (*Token, error) {
+	tokenHash := hashToken(plaintextToken)
+	var token Token
+	err := t.db.Get(
+		&token,
+		`SELECT token_encrypted, user_id, device_id, last_seen FROM syncv3_sync2_tokens WHERE token_hash=$1`,
+		tokenHash,
+	)
+	if err != nil {
+		return nil, err
+	}
+	token.AccessToken = plaintextToken
+	token.AccessTokenHash = tokenHash
+	return &token, nil
+}
+
+type TokenWithSince struct {
```

Comment:
Bad name as it is awkward when we want to add additional fields to this struct beyond "since".

Reply:
I wasn't happy with it (but also I find it relatively easy to rename things---IntelliJ's/Goland's refactoring tools seem to work very well for go.